jupyter_client/multikernelmanager.py

```diff
@@ -215,15 +256,31 @@ async def _async_shutdown_kernel(
             Will the kernel be restarted?
         """
         self.log.info("Kernel shutdown: %s" % kernel_id)
-        if kernel_id in self._starting_kernels:
+        # If we're using pending kernels, block shutdown when a kernel is pending.
+        if self._using_pending_kernels() and kernel_id in self._pending_kernels:
+            raise RuntimeError("Kernel is in a pending state. Cannot shutdown.")
+        # If the kernel is still starting, wait for it to be ready.
+        elif kernel_id in self._starting_kernels:
```

Comment:
Is the use of `_starting_kernels` intentional?  Technically, this could also contain kernels that are shutting down now, but only when pending kernels are enabled - so I suspect this was for hinting that, in this case, the kernel can only be starting.

Perhaps we could be more explicit (and save an existence check [nit]) via: 
```python
        if kernel_id in self._pending_kernels:
            if self._using_pending_kernels():
                 raise RuntimeError("Kernel is in a pending state. Cannot shutdown.")
            else:  # kernel is still starting, wait for its startup
                kernel = self._pending_kernels[kernel_id]
                try:
                    await kernel
                except Exception:
                    self.remove_kernel(kernel_id)
```